### PR TITLE
Software type plugins can only expose exactly one software type

### PR DESCRIPTION
**Pull request by ghale**
This changes the software type plugin validation to only allow plugins to expose exactly one software type.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by ghale** [comment]
@bot-gradle test this

File: platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy [hunk 1]
```diff
@@ -263,7 +239,33 @@ class SoftwareTypeDeclarationIntegrationTest extends AbstractIntegrationSpec imp
...
+            }
+        """
+
+        when:
+        fails(":printTestSoftwareTypeExtensionConfiguration")
+
+        then:
+        failure.assertHasCause("Failed to apply plugin 'com.example.test-software-type'.")
+        failure.assertHasCause("A problem was found with the SoftwareTypeImplPlugin plugin.")
+        failure.assertHasCause("Type 'org.gradle.test.SoftwareTypeImplPlugin' is registered as a software type plugin, but it exposes multiple software types.")
```
Comment (tresat) [inline platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy:268]: 🤔 Can we expand this message to note the names of software types declared?  Something like:

```
Plugin with type 'com.example.restricted.RestrictedPlugin' is registered as a software type plugin, but it declares more than one software type: application, library, utility. Software type plugins must expose only a single software type.
```

Alternately, as the exception already implements `ResolutionProvider`, we could add a suggestion to expose only a single software type or register multiple plugins.

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java [hunk 2]
```diff
@@ -81,23 +87,56 @@ private void registerSoftwareTypes(TypeMetadata typeMetadata) {
...
         TypeToken<?> softwareTypePluginImplType = TypeToken.of(softwareTypePluginImplClass);
         TypeMetadata softwareTypePluginImplMetadata = inspectionScheme.getMetadataStore().getTypeMetadata(softwareTypePluginImplType.getRawType());
+        softwareTypePluginImplMetadata.visitValidationFailures(null, typeValidationContext);
 
-        for (PropertyMetadata propertyMetadata : softwareTypePluginImplMetadata.getPropertiesMetadata()) {
-            Optional<SoftwareType> registersSoftwareType = propertyMetadata.getAnnotation(SoftwareType.class);
-            if (registersSoftwareType.isPresent()) {
-                return;
-            }
+        long exposedSoftwareTypes = softwareTypePluginImplMetadata.getPropertiesMetadata().stream()
```
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java:102]: 💭  
```
List<String> registeredSoftwareTypes = softwareTypePluginImplMetadata.getPropertiesMetadata().stream()
            .map(propertyMetadata -> propertyMetadata.getAnnotation(SoftwareType.class))
            .filter(Optional::isPresent)
            .map(annotation -> annotation.get().name())
            .sorted()
            .collect(Collectors.toList());
```

And then have the list of names to put in the error message, which can also be counted.

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java [hunk 3]
```diff
@@ -81,23 +87,56 @@ private void registerSoftwareTypes(TypeMetadata typeMetadata) {
...
+                    .contextualLabel("is registered as a software type plugin, but it exposes multiple software types")
+                    .severity(Severity.ERROR)
+                    .details("This class was registered as a software type plugin, but it exposes multiple software types. Software type plugins must expose exactly one software type via a property with the @SoftwareType annotation.")
+                    .solution("Add the @SoftwareType annotation to only one property of " + softwareTypePluginImplClass.getSimpleName())
+            );
         }
 
-        throw new InvalidUserDataException("A plugin with type '" + softwareTypePluginImplClass.getName() + "' was registered as a software type plugin, but it does not expose any software types. Software type plugins must expose software types via properties with the @SoftwareType annotation.");
+        if (!typeValidationContext.getProblems().isEmpty()) {
+            throw new DefaultMultiCauseException(
```
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java:129]: 🤔 Can't you only ever get a single validation problem out of this code?  Why start with `MultiCause`?

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java [hunk 4]
```diff
@@ -81,23 +87,56 @@ private void registerSoftwareTypes(TypeMetadata typeMetadata) {
...
+                    .contextualLabel("is registered as a software type plugin but does not expose a software type")
+                    .severity(Severity.ERROR)
+                    .details("This class was registered as a software type plugin, but it does not expose a software type. Software type plugins must expose exactly one software type via a property with the @SoftwareType annotation.")
+                    .solution("Add @SoftwareType annotations to properties of " + softwareTypePluginImplClass.getSimpleName())
+                    .solution("Remove " + softwareTypePluginImplClass.getSimpleName() + " from the @RegistersSoftwareTypes annotation on " + registeringPlugin.getSimpleName())
+            );
+        } else if (exposedSoftwareTypes > 1) {
+            typeValidationContext.visitTypeProblem(problem ->
+                problem.withAnnotationType(softwareTypePluginImplClass)
+                    .id("missing-software-type", "Missing software type annotation", GradleCoreProblemGroup.validation().type())
```
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java:120]: ❕ I think this needs a different id - a type is not missing here, so this error should be distinct from the missing case.

File: platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy [hunk 5]
```diff
@@ -263,7 +239,33 @@ class SoftwareTypeDeclarationIntegrationTest extends AbstractIntegrationSpec imp
 
         then:
         failure.assertHasCause("Failed to apply plugin 'com.example.test-software-type'.")
-        failure.assertHasCause("A plugin with type 'org.gradle.test.SoftwareTypeImplPlugin' was registered as a software type plugin, but it does not expose any software types. Software type plugins must expose software types via properties with the @SoftwareType annotation.")
+        failure.assertHasCause("A problem was found with the SoftwareTypeImplPlugin plugin.")
+        failure.assertHasCause("Type 'org.gradle.test.SoftwareTypeImplPlugin' is registered as a software type plugin but does not expose a software type.")
+    }
+
+    def 'sensible error when a software type plugin is registered that exposes multiple software types'() {
```
Comment (tresat) [inline platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy:246]: 💭 Probably also should have a case ensuring you can register multiple software types in multiple plugins using single RegistersSoftwareTypes plugin.

**Review by tresat — APPROVED** [review]
I didn't realize you were working on this, I was just about to submit my own PR with a very similar change.

"Call it in the air!" 😄

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java (hunk 3 above)
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java:129]: A few lines up, we call `softwareTypePluginImplMetadata.visitValidationFailures()` which captures any validation failures it encountered when inspecting the type.  Right now this only includes the validations in `SoftwareTypeAnnotationHandler` which checks that the public type is a super class of (or the same class as) the method return type.  So you could have say, mismatched types _and_ multiple exposed types.

File: platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy (hunk 5 above)
Comment (ghale) [reply platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy:246]: There already tests for this [here](https://github.com/gradle/gradle/blob/e60c047483af8dd39fc28715556282dbe1f24dd7/platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy#L110) and [here](https://github.com/gradle/gradle/blob/e60c047483af8dd39fc28715556282dbe1f24dd7/platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy#L138).

File: platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy (hunk 1 above)
Comment (ghale) [reply platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeDeclarationIntegrationTest.groovy:268]: Yep.  I think that's a little better.

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java (hunk 2 above)
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java]: Done.

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java (hunk 4 above)
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java]: Good catch.  Copy-and-paste error.

**Comment by ghale** [comment]
Sorry for not coordinating, @tresat.  I got blocked on other stuff, and this was fresh in my mind, so I just banged it out.  Should have at least given you a heads up.  Thanks for the good suggestions.

**Comment by ghale** [comment]
@bot-gradle test this

**Pull request merged**